staking/programs/staking/src/state/positions.rs

```diff
@@ -3,59 +3,156 @@ use {
 
-/// An array that contains all of a user's positions i.e. where are the staking and who are they
-/// staking to.
-/// The invariant we preserve is : For i < next_index, positions[i] == Some
-/// For i >= next_index, positions[i] == None
-
+/// The header of DynamicPositionArray
 #[account(zero_copy)]
 #[repr(C)]
 pub struct PositionData {
     pub owner: Pubkey,
-    positions: [[u8; POSITION_BUFFER_SIZE]; MAX_POSITIONS],
 }
 
 impl PositionData {
-    pub const LEN: usize = 8 + 32 + MAX_POSITIONS * POSITION_BUFFER_SIZE;
+    pub const LEN: usize = 8 + 32;
 }
 
-impl Default for PositionData {
-    // Only used for testing, so unwrap is acceptable
-    fn default() -> Self {
-        PositionData {
-            owner:     Pubkey::default(),
-            positions: [[0u8; POSITION_BUFFER_SIZE]; MAX_POSITIONS],
+/// This account stores a user's positions in a dynamic sized array.
+/// Its first 40 bytes are `PositionData` (including discriminator) and the rest is a
+/// variable-length slice of `[u8; POSITION_BUFFER_SIZE]`. Each element of the array can be
+/// deserialized into an `Option<Position>`. The old invariant is maintained: For `i < next_index`,
+/// `positions[i] == Some` For `i >= next_index`, `positions[i] == None`
+/// Other invariants are that `data_len() == 40 + n * POSITION_BUFFER_SIZE` where n is an integer
+/// and that `data_len() >= 40 + next_index * POSITION_BUFFER_SIZE`
+pub struct DynamicPositionArray<'a> {
+    pub acc_info: AccountInfo<'a>, /* We store account info to get access to the data and
+                                    * resize */
+}
+
+impl<'a> DynamicPositionArray<'a> {
+    fn get_positions_slice(&self) -> Result<&mut [[u8; POSITION_BUFFER_SIZE]]> {
+        let position_capacity = self.get_position_capacity();
+        unsafe {
+            Ok(std::slice::from_raw_parts_mut(
+                self.acc_info.try_borrow_mut_data()?[PositionData::LEN..].as_mut_ptr()
+                    as *mut [u8; POSITION_BUFFER_SIZE],
+                position_capacity,
+            ))
         }
     }
-}
-impl PositionData {
-    pub fn initialize(&mut self, owner: &Pubkey) {
-        self.owner = *owner;
+
+    fn data_len(&self) -> usize {
+        self.acc_info.data_len()
+    }
+
+    pub fn load_init(account_loader: &AccountLoader<'a, PositionData>) -> Result<Self> {
+        let acc_info = account_loader.to_account_info();
+        if !acc_info.is_writable {
+            return Err(AnchorErrorCode::AccountNotMutable.into());
+        }
+
+        {
+            let data = acc_info.try_borrow_mut_data()?;
+
+            // The discriminator should be zero, since we're initializing.
+            let mut disc_bytes = [0u8; 8];
+            disc_bytes.copy_from_slice(&data[..8]);
+            let discriminator = u64::from_le_bytes(disc_bytes);
+            if discriminator != 0 {
+                return Err(AnchorErrorCode::AccountDiscriminatorAlreadySet.into());
+            }
+        }
```

Comment:
what was the reason to wrap this in `{}`?

Reply:
It borrows `account_info` so I'd have clone account_info below